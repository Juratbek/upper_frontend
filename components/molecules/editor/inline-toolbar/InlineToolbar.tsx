import { ButtonHTMLAttributes, useState } from 'react';
import { Override } from 'utils';

import { useEditorContext } from '../context';
import { INLINE_TOOLS } from './InlineToolbar.constants';
import cls from './InlineToolbar.module.scss';
import { IInlineTool } from './InlineToolbar.types';

export const InlineToolbar = () => {
  const { inlineToolbar, hideInlineToolbar } = useEditorContext();
  const [activeTool, setActiveTool] = useState<IInlineTool>();

  const itemClickHandler = (tool: IInlineTool) => {
    tool.callback?.();

    setActiveTool((prevTool) => {
      if (prevTool) return undefined;
      return tool;
    });
  };

  const closePopover = () => setActiveTool(undefined);

  return (
    <div
      className={cls.root}
      id='inline-toolbar'
      style={inlineToolbar.position ? { ...inlineToolbar.position } : { display: 'none' }}
    >
      <div className={cls['tools']}>
        {INLINE_TOOLS.map((tool, index) => (
          <Item onClick={() => itemClickHandler(tool)} key={index}>
            {tool.icon}
          </Item>
        ))}
      </div>
      {activeTool?.renderPopover && (
        <div className={cls.popover}>
          {activeTool.renderPopover({ close: closePopover, closeToolbar: hideInlineToolbar })}
        </div>
      )}
    </div>
  );
};

const Item = ({
  children,
  ...props
}: Override<ButtonHTMLAttributes<HTMLButtonElement>, { children: string }>) => {
  return <button {...props} className={cls.item} dangerouslySetInnerHTML={{ __html: children }} />;
};
