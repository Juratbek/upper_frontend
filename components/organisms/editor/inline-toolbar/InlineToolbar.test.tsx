import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';

import { InlineToolbar } from './InlineToolbar';

const mocks = vi.hoisted(() => ({
  useEditorContext: vi.fn().mockReturnValue({
    inlineToolbar: {},
    hideInlineToolbar: vi.fn(),
  }),
  inlineToolCallback: vi.fn(),
}));

vi.mock('./InlineToolbar.constants', async (importOriginal) => {
  const original = await importOriginal<typeof import('./InlineToolbar.constants')>();

  const INLINE_TOOLS = original.INLINE_TOOLS.map((tool) => {
    tool.callback = mocks.inlineToolCallback;
    return tool;
  });

  return { ...original, INLINE_TOOLS };
});

vi.mock('../context', () => ({
  useEditorContext: mocks.useEditorContext,
}));

describe('Inline toolbar', () => {
  afterEach(() => {
    vi.resetAllMocks();
  });

  it('snapshot', () => {
    render(<InlineToolbar />);
    expect(document.body).toMatchSnapshot();
  });

  it('snapshot with positions', () => {
    mocks.useEditorContext.mockReturnValue({
      inlineToolbar: {
        position: { top: 10, left: 10 },
      },
      hideInlineToolbar: vi.fn(),
    });
    render(<InlineToolbar />);

    expect(document.body).toMatchSnapshot();
  });

  it('intine tools click handler', async () => {
    mocks.useEditorContext.mockReturnValue({
      inlineToolbar: {
        position: { top: 10, left: 10 },
      },
      hideInlineToolbar: vi.fn(),
    });
    render(<InlineToolbar />);

    const buttons = screen.getAllByRole<HTMLButtonElement>('button');

    buttons.forEach((button) => {
      userEvent.click(button);
    });

    await waitFor(() => {
      expect(mocks.inlineToolCallback).toHaveBeenCalledTimes(buttons.length);
    });
  });

  it('opens inline toolbar popover on click', async () => {
    const { INLINE_TOOLS } = await import('./InlineToolbar.constants');

    mocks.useEditorContext.mockReturnValue({
      inlineToolbar: {
        position: { top: 10, left: 10 },
      },
      hideInlineToolbar: vi.fn(),
    });
    render(<InlineToolbar />);

    const buttons = screen.getAllByRole<HTMLButtonElement>('button');
    const toolWithPopoverIndex = INLINE_TOOLS.findIndex(
      (tool) => typeof tool.renderPopover === 'function',
    );
    const linkTool = buttons[toolWithPopoverIndex];

    userEvent.click(linkTool);

    await waitFor(() => {
      expect(mocks.inlineToolCallback).toHaveBeenCalledTimes(1);
      expect(document.body).toMatchSnapshot();
    });
  });
});
