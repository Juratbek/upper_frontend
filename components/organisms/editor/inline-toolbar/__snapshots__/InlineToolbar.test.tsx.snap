// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Inline toolbar > opens inline toolbar popover on click 1`] = `
<body>
  <div>
    <div
      class="root"
      id="inline-toolbar"
      style="top: 10px; left: 10px;"
    >
      <div
        class="tools"
      >
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 12L9 7.1C9 7.04477 9.04477 7 9.1 7H10.4C11.5 7 14 7.1 14 9.5C14 9.5 14 12 11 12M9 12V16.8C9 16.9105 9.08954 17 9.2 17H12.5C14 17 15 16 15 14.5C15 11.7046 11 12 11 12M9 12H11"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13.34 10C12.4223 12.7337 11 17 11 17"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
            <path
              d="M14.21 7H14.2"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.69998 12.6L7.67896 12.62C6.53993 13.7048 6.52012 15.5155 7.63516 16.625V16.625C8.72293 17.7073 10.4799 17.7102 11.5712 16.6314L13.0263 15.193C14.0703 14.1609 14.2141 12.525 13.3662 11.3266L13.22 11.12"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
            <path
              d="M16.22 11.12L16.3564 10.9805C17.2895 10.0265 17.3478 8.5207 16.4914 7.49733V7.49733C15.5691 6.39509 13.9269 6.25143 12.8271 7.17675L11.3901 8.38588C10.0935 9.47674 9.95706 11.4241 11.0888 12.6852L11.12 12.72"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 8L5 12L9 16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M15 8L19 12L15 16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
      <div
        class="popover"
        role="dialog"
      >
        <input
          class="input"
          placeholder="Havolani kiriting"
          value=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`Inline toolbar > opens inline toolbar popover on click 2`] = `
<body>
  <div>
    <div
      class="root"
      id="inline-toolbar"
      style="top: 10px; left: 10px;"
    >
      <div
        class="tools"
      >
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 12L9 7.1C9 7.04477 9.04477 7 9.1 7H10.4C11.5 7 14 7.1 14 9.5C14 9.5 14 12 11 12M9 12V16.8C9 16.9105 9.08954 17 9.2 17H12.5C14 17 15 16 15 14.5C15 11.7046 11 12 11 12M9 12H11"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13.34 10C12.4223 12.7337 11 17 11 17"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
            <path
              d="M14.21 7H14.2"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.69998 12.6L7.67896 12.62C6.53993 13.7048 6.52012 15.5155 7.63516 16.625V16.625C8.72293 17.7073 10.4799 17.7102 11.5712 16.6314L13.0263 15.193C14.0703 14.1609 14.2141 12.525 13.3662 11.3266L13.22 11.12"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
            <path
              d="M16.22 11.12L16.3564 10.9805C17.2895 10.0265 17.3478 8.5207 16.4914 7.49733V7.49733C15.5691 6.39509 13.9269 6.25143 12.8271 7.17675L11.3901 8.38588C10.0935 9.47674 9.95706 11.4241 11.0888 12.6852L11.12 12.72"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 8L5 12L9 16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M15 8L19 12L15 16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
      <div
        class="popover"
        role="dialog"
      >
        <input
          class="input"
          placeholder="Havolani kiriting"
          value=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`Inline toolbar > snapshot 1`] = `
<body>
  <div>
    <div
      class="root"
      id="inline-toolbar"
      style="display: none;"
    >
      <div
        class="tools"
      >
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 12L9 7.1C9 7.04477 9.04477 7 9.1 7H10.4C11.5 7 14 7.1 14 9.5C14 9.5 14 12 11 12M9 12V16.8C9 16.9105 9.08954 17 9.2 17H12.5C14 17 15 16 15 14.5C15 11.7046 11 12 11 12M9 12H11"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13.34 10C12.4223 12.7337 11 17 11 17"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
            <path
              d="M14.21 7H14.2"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.69998 12.6L7.67896 12.62C6.53993 13.7048 6.52012 15.5155 7.63516 16.625V16.625C8.72293 17.7073 10.4799 17.7102 11.5712 16.6314L13.0263 15.193C14.0703 14.1609 14.2141 12.525 13.3662 11.3266L13.22 11.12"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
            <path
              d="M16.22 11.12L16.3564 10.9805C17.2895 10.0265 17.3478 8.5207 16.4914 7.49733V7.49733C15.5691 6.39509 13.9269 6.25143 12.8271 7.17675L11.3901 8.38588C10.0935 9.47674 9.95706 11.4241 11.0888 12.6852L11.12 12.72"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 8L5 12L9 16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M15 8L19 12L15 16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`Inline toolbar > snapshot with positions 1`] = `
<body>
  <div>
    <div
      class="root"
      id="inline-toolbar"
      style="top: 10px; left: 10px;"
    >
      <div
        class="tools"
      >
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 12L9 7.1C9 7.04477 9.04477 7 9.1 7H10.4C11.5 7 14 7.1 14 9.5C14 9.5 14 12 11 12M9 12V16.8C9 16.9105 9.08954 17 9.2 17H12.5C14 17 15 16 15 14.5C15 11.7046 11 12 11 12M9 12H11"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13.34 10C12.4223 12.7337 11 17 11 17"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
            <path
              d="M14.21 7H14.2"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.69998 12.6L7.67896 12.62C6.53993 13.7048 6.52012 15.5155 7.63516 16.625V16.625C8.72293 17.7073 10.4799 17.7102 11.5712 16.6314L13.0263 15.193C14.0703 14.1609 14.2141 12.525 13.3662 11.3266L13.22 11.12"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
            <path
              d="M16.22 11.12L16.3564 10.9805C17.2895 10.0265 17.3478 8.5207 16.4914 7.49733V7.49733C15.5691 6.39509 13.9269 6.25143 12.8271 7.17675L11.3901 8.38588C10.0935 9.47674 9.95706 11.4241 11.0888 12.6852L11.12 12.72"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <button
          class="item"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 8L5 12L9 16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M15 8L19 12L15 16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</body>
`;
